### ISSUE TITLE
Rest parameters don't work with generators

### ISSUE THREAD

**Issue by YourDeveloperFriend** [6to5/6to5#438]
There's a bug where a variable _ref isn't defined when you declare a generator using a spread.

ES6:
```Javascript
let a = function*(args) {
  this.runRoute(...args);
};
```

Compiled:
```Javascript
"use strict";

var _toArray = function (arr) {
  return Array.isArray(arr) ? arr : Array.from(arr);
};

var a = regeneratorRuntime.mark(function callee$0$0(args) {
  return regeneratorRuntime.wrap(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        (_ref = this).runRoute.apply(_ref, _toArray(args)); // _ref is not defined
      case 1:
      case "end":
        return context$1$0.stop();
    }
  }, callee$0$0, this);
});
```

You can [check it out](http://6to5.org/repl/#?experimental=true&playground=true&evaluate=true&code=%0Alet%20a%20%3D%20function*(args)%20%7B%0A%20%20this.runRoute(...args)%3B%0A%7D%3B%0A)

I think the requirements to duplicate this bug are the following:
- The function must be a generator.
- The function must use a spread.
- The function must be calling another function on `this`.

You really just need to add a `let _ref = this;` sometime before.

For now, I'll just have to write the spread manually:

```Javascript
let a = function*(args) {
  this.runRoute.apply(this, args);
};
```

**Comment by sebmck** [comment]
Looks like a regenerator issue. I'll investigate shortly.

**Comment by sebmck** [comment]
Related: #338.

@benjamn These are the type of issues that were resolved by embedding regenerator. 6to5 stores some hidden properties on nodes that get replaced by regenerator. Any ideas?

**Comment by benjamingr** [comment]
@sebmck  I wouldn't say this is a regenerator issue since it does not break anything in regenerator nor the guarantees it makes as far as I understand.

I'm just wondering here - can't you sweep the nodes and store the hidden properties before invoking regenerator on the code, then swap them back in?

**Comment by sebmck** [comment]
@benjamingr I was more referring to the 6to5 regenerator integration, not necessarily regenerator itself.

> I'm just wondering here - can't you sweep the nodes and store the hidden properties before invoking regenerator on the code, then swap them back in?

We can't determine what nodes have been replaced, or what their source node could even possibly be.

**Comment by benjamingr** [comment]
@sebmck I don't know how documented this is but from playing with regenerator it is pretty consistent with how it partitions the `yield` statements into cases. That's probably not the best approach though.

I'm guessing here and it doesn't really align with the current architecture - but the information of what moved where is contained in the source maps - can't 6to5 use those to figure out how things moved around?

**Comment by sebmck** [comment]
@benjamingr Not really, if it is then it'd be a massive pain. Just seems like a major hassle when there could possibly be some API for it.

**Comment by benjamingr** [comment]
Oh I completely agree that it would be a huge hassle - I'm just wondering if you're not expecting regenerator here to support something that requires internal understanding of 6to5...

How would an optimal solution look in your opinion given regenerator replacing nodes is fundamental and can't be changed?

**Comment by sebmck** [comment]
@benjamingr Best-case would be for there to be provide regenerator a callback that it'll call with `(oldNode, newNode)`.

**Comment by kevinb7** [comment]
I ran this code through the REPL:
```
let b = function(args) {
  this.runRoute(...args);
};

var c = function*(args) {
  var _ref2;
  (_ref2 = this).runRoute.apply(_ref2, _toArray(args));
};
```
I wanted to see what the spread transformed look like on a normal function.  I manually applied that transform to a generator function and the result is ```c```.  Below are the transformed versions of ```b``` and ```c```.
```
var b = function (args) {
  var _ref;
  (_ref = this).runRoute.apply(_ref, _toArray(args));
};

var c = regeneratorRuntime.mark(function callee$0$0(args) {
  var _ref2;
  return regeneratorRuntime.wrap(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        (_ref2 = this).runRoute.apply(_ref2, _toArray(args));
      case 1:
      case "end":
        return context$1$0.stop();
    }
  }, callee$0$0, this);
});
```
It appears that ```c``` is what we would expect for output.  This leads me to believe that the spread transform is being run after regenerator is run.  If this is the case, I would expect to see something like the following in the code that @YourDeveloperFriend posted:
```
case 0:
        var _ref = this;
        (_ref = this).runRoute.apply(_ref, _toArray(args));
```
I don't know what's cause the missing "_ref = this;", but if regenerator is run beforehand it doesn't really matter because it's probably going to end up in the wrong place because the spread transform doesn't know what regenerator is doing.

TL;DR: It seems like running the spread transform first and regenerator last might solve this issue.  I need to grok the code a bit and then I'll try making this change.

**Comment by sebmck** [comment]
@kevinb7 I already know what the cause of it is. We store `_declarations` on scope nodes so we can dynamically push variables to them. This is especially useful for when we need to assign to a variable as we need a declaration to do that to. Regenerator replaces nodes and we lose the dynamic declarations that we've set on the scope.

**Comment by monsanto** [comment]
We should re-inline Regenerator for 3.0.0.

**Comment by benjamn** [comment]
Making sure different transforms respect each other's hidden assumptions is a tricky problem in general. When possible, it's usually better to make concrete changes to the AST itself, rather than relying on "private" metadata and hard-to-enforce invariants, but I know you wouldn't rely on those things if you had a choice. That said, maybe we could decide on some standard way of storing node metadata, so that AST transforms could preserve that information when they create new nodes to replace old nodes?

**Comment by sebmck** [comment]
@benjamn At the moment I think it's easier to just maintain a fork. I've created [regenerator-6to5](https://github.com/6to5/regenerator-6to5/compare/facebook:master...master) that just removes recast, uses `leadingComments` and `trailingComments` instead of `comments`, uses 6to5's block hoisting instead of `ast-types`. I hope you don't mind, I'll regularly pull from upstream with the goal of eventually using vanilla `regenerator` once certain changes are made.

**Comment by benjamn** [comment]
I'm actually pretty happy with that. As long as it's a fork in the GitHub sense, pull requests and rebases should be easy enough.

**Comment by sebmck** [comment]
Fixed via da8888a3507ebd19e1d5dd3391340747b21d30f8 in the 3.0.0 branch.

**Comment by kevinb7** [comment]
nice!

**Comment by YourDeveloperFriend** [comment]
Thanks!

**Comment by sebmck** [comment]
Fixed as of 3.0.0
